src/GitVersionTask/GenerateGitVersionInformation.cs

```diff
@@ -1,42 +1,99 @@
+            )
+        {
+            if ( !input.ValidateInput() )
+            {
+                throw new Exception( "Invalid input." );
+            }
 
-        [Required]
-        public string IntermediateOutputPath { get; set; }
+            var logger = new TaskLogger();
+            Logger.SetLoggers( logger.LogInfo, logger.LogInfo, logger.LogWarning, s => logger.LogError( s ) );
 
-        [Required]
-        public string Language { get; set; }
 
-        [Output]
-        public string GitVersionInformationFilePath { get; set; }
+            Output output = null;
+            try
+            {
+                output = InnerExecute(input );
+            }
+            catch (WarningException errorException)
+            {
+                logger.LogWarning(errorException.Message);
+                output = new Output();
+            }
+            catch (Exception exception)
+            {
+                logger.LogError("Error occurred: " + exception);
+                throw;
+            }
+            finally
+            {
+                Logger.Reset();
+            }
 
-        protected override void InnerExecute()
+            return output;
+        }
+
+        private static Output InnerExecute( Input input )
         {
-            if (GetVersionVariables(out var versionVariables)) return;
+            var execute = GitVersionTaskBase.CreateExecuteCore();
+            if (!execute.TryGetVersion(input.SolutionDirectory, out var versionVariables, input.NoFetch, new Authentication()))
+            {
+                return null;
+            }
 
-            var fileExtension = TaskUtils.GetFileExtension(Language);
-            var fileName = $"GitVersionInformation.g.{fileExtension}";
+            var fileWriteInfo = input.IntermediateOutputPath.GetWorkingDirectoryAndFileNameAndExtension(
+                input.Language,
+                input.ProjectFile,
+                ( pf, ext ) => $"GitVersionInformation.g.{ext}",
+                ( pf, ext ) => $"GitVersionInformation_{Path.GetFileNameWithoutExtension( pf )}_{Path.GetRandomFileName()}.g.{ext}"
+                );
 
-            if (IntermediateOutputPath == null)
+            var output = new Output()
             {
-                fileName = $"GitVersionInformation_{Path.GetFileNameWithoutExtension(ProjectFile)}_{Path.GetRandomFileName()}.g.{fileExtension}";
-            }
+                GitVersionInformationFilePath = Path.Combine( fileWriteInfo.WorkingDirectory, fileWriteInfo.FileName )
+            };
+            var generator = new GitVersionInformationGenerator( fileWriteInfo.FileName, fileWriteInfo.WorkingDirectory, versionVariables, new FileSystem());
+            generator.Generate();
 
-            var workingDirectory = IntermediateOutputPath ?? TempFileTracker.TempPath;
+            return output;
+        }
 
-            GitVersionInformationFilePath = Path.Combine(workingDirectory, fileName);
+        public sealed class Input
```

Comment:
@stazz, why isn't any form of inheritance used across the different `Input` classes (and `Output` classes, for that matter)?

Reply:
I simply forgot to do that. I'll fix this on Sunday.

Reply:
Great!